main/src/library/Chain.flix

```diff
@@ -0,0 +1,847 @@
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def findRight(f: a -> Bool, xs: Chain[a]): Option[a] = match viewRight(xs) {
+        case NoneRight        => None
+        case SomeRight(rs, x) => if (f(x)) Some(x) else findRight(f, rs)
+    }
+
+    ///
+    /// Returns a list of all integers between `b` (inclusive) and `e` (exclusive).
+    ///
+    /// Returns `Nil` if `b >= e`.
+    ///
+    pub def range(b: Int32, e: Int32): Chain[Int32] = rangeHelper(b, e, Chain.empty())
+
+    ///
+    /// Helper function for `range`.
+    ///
+    def rangeHelper(b: Int32, e: Int32, ac: Chain[Int32]): Chain[Int32] =
+        if (b >= e) ac else rangeHelper(b + 1, e, snoc(ac, b))
+
+    ///
+    /// Returns a list with the element `a` repeated `n` times.
+    ///
+    /// Returns `Nil` if `n < 0`.
+    ///
+    pub def repeat(n: Int32, a: a): Chain[a] = repeatHelper(n, a, Empty)
+
+    ///
+    /// Helper function for `repeat`.
+    ///
+    def repeatHelper(n: Int32, a: a, ac: Chain[a]): Chain[a] =
+        if (n <= 0)
+            ac
+        else {
+            let ac1 = cons(a, ac);
+            repeatHelper(n-1, a, ac1)
+        }
+
+    ///
+    /// Alias for `scanLeft`.
+    ///
+    pub def scan(f: (b, a) -> b & ef, s: b, xs: Chain[a]): Chain[b] & ef = scanLeft(f, s, xs)
+
+    ///
+    /// Accumulates the result of applying `f` to `xs` going left to right.
+    ///
+    /// That is, the result is of the form: `s :: f(s, x1) :: f(f(s, x1), x2)  ...`.
+    ///
+    pub def scanLeft(f: (b, a) -> b & ef, s: b, xs: Chain[a]): Chain[b] & ef = scanLeftHelper(f, s, xs, point(s))
+
+    ///
+    /// Helper function for `scanLeft`.
+    ///
+    def scanLeftHelper(f: (b, a) -> b & ef, s: b, xs: Chain[a], ac: Chain[b]): Chain[b] & ef = match viewLeft(xs) {
+        case NoneLeft        => ac
+        case SomeLeft(x, rs) => {let a = f(s, x); scanLeftHelper(f, a, rs, snoc(ac, a))}
+    }
+
+    ///
+    /// Accumulates the result of applying `f` to `xs` going right to left.
+    ///
+    /// That is, the result is of the form: `... f(xn-1, f(xn, s)) :: f(xn, s) :: s`.
+    ///
+    pub def scanRight(f: (a, b) -> b & ef, s: b, xs: Chain[a]): Chain[b] & ef = scanRightHelper(f, s, xs, point(s))
+
+    ///
+    /// Helper function for `scanLeft`.
+    ///
+    def scanRightHelper(f: (a, b) -> b & ef, s: b, xs: Chain[a], ac: Chain[b]): Chain[b] & ef = match viewRight(xs) {
+        case NoneRight        => ac
+        case SomeRight(rs, x) => {let a = f(x, s); scanRightHelper(f, a, rs, cons(a, ac))}
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs`.
+    ///
+    /// That is, the result is of the form: `f(x1) :: f(x2) :: ...`.
+    ///
+    pub def map(f: a -> b & ef, xs: Chain[a]): Chain[b] & ef =
+        mapHelper(f, xs, Empty)
+
+    ///
+    /// Helper function for `map`.
+    ///
+    def mapHelper(f: a -> b & ef, xs: Chain[a], ac: Chain[b]): Chain[b] & ef = match viewLeft(xs) {
```

Comment:
A1 - `map` visits the chain left to right.

A2 - the chain has the same structure after map.

`map` could flatten the chain out with the expectation that further processing is most usually left-to-right (the library already has a bit of left-to-right bias e.g. `count`), I've a "natural" inclination against this but I don't have evidence to prove a case. 

I see Chain as a transitional data structure that's lifetime should be as long as it's being constructed but if there is a natural point at which it would be analysed (mapped, counted, ...) and not built further, it should be transformed to a list or array.

Reply:
OK, and thanks for the clarification. I agree with your view. Until evidence emerges to the contrary let us keep it as is.